### PR TITLE
validation: Add missing mempool locks

### PR DESCRIPTION
**Pull request by maflcko**
Take the mempool read lock during reorgs, so that we don't accidentally read an inconsistent mempool.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2686,6 +2686,7 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
 
         {
             LOCK(cs_main);
+            LOCK(::mempool.cs);
```
Comment (practicalswift) [inline src/validation.cpp]: Nit: `LOCK2(cs_main, ::mempool.cs);`

**Comment by practicalswift** [comment]
AppVeyor builds fine, but no Travis checking for this PR. Travis integration broken?

File: src/validation.cpp (hunk 1 above)
Comment (ryanofsky) [inline src/validation.cpp]: What's the reason for acquiring mempool lock at a broad scope instead of more narrowly where the mempool is actually accessed? Would be nice to have comments in the places where this PR adds locking to indicate purpose of locks and reasoning behind placement.

**Comment by sdaftuar** [comment]
Is the idea that you're trying to prevent RPC users from having an inconsistent state, like if they call getbestblockhash() and then call getrawmempool(), the latter should be consistent with the former?

That may be reasonable, but I think it'd be good to have some documentation (whether a project document or even a gist explaining our thinking) explaining the consistency guarantees for RPC calls with respect to the mempool, validation, and the wallet.  I feel like this kind of thing has come up before in a haphazard way and caused issues when we didn't think things through or missed an implication of a change.

**Comment by ryanofsky** [comment]
> Is the idea that you're trying to prevent RPC users from having an inconsistent state

From the description, I assumed the intent of the PR was to prevent RPCs like `getmempoolentry`, which acquire `mempool.cs` but not `cs_main`, from returning garbage or segfaulting during reorgs because `UpdateMempoolForReorg` and `InvalidateBlock` functions were apparently not acquiring `mempool.cs` while updating the mempool.

But actually it seems like in all the places where the mempool is updated, the code already has both `cs_main` and `mempool.cs` locks. So it should be perfectly safe to read the mempool while holding either of the two locks, and trying to acquire both is wasteful.

So maybe this PR is not doing what it intended to do, and instead it would be better just to add comments and fix thread safety annotations somehow. Maybe it would be possible to annotate `cs_main` and `mempool.cs` individually as shared locks needed for reading mempool data, and `cs_main`+`mempool.cs` both together as an exclusive lock needed for updating mempool data?

**Comment by MarcoFalke** [comment]
@ryanofsky With inconsistent I don't mean invalid reads, but rather transactions that are invalid (and still in the process of removal) as of our current block tip.

**Review by ryanofsky — COMMENTED** [review]
Code utACK fa3e0471a81252e333886b97d8d33bb275061d42. Code change seems perfectly fine, and it could perhaps lead to more consistency in some cases, but like the previous comment https://github.com/bitcoin/bitcoin/pull/14193#issuecomment-421375559 says, it's unclear which cases are better off after this change or what case motivated this PR. More comments and documentation would be helpful.

**Comment by MarcoFalke** [comment]
@sdaftuar, @ryanofsky: Added a test with comments to illustrate my motivation a bit more.

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -468,21 +468,29 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
+        // Release lock while waiting
```
Comment (ryanofsky) [inline src/rpc/mining.cpp]: In commit "validation: Add missing mempool locks" (dea5dbf2d5ca0177d3a4be665790288bef40e27e)

This is hard to follow. Can you mention this change the commit description? It seems like there is no difference in behavior here except to avoid holding `g_best_block_mutex` while calling `mempool.GetTransactionsUpdated()`? Or am I misinterpreting the diff? If that is the change, it seems good to release the lock, but why is it important do to that now when it wasn't needed before?

**Review by ryanofsky — COMMENTED** [review]
Started review:
- [X] dea5dbf2d5ca0177d3a4be665790288bef40e27e validation: Add missing mempool locks (1/2)
- [ ] fa27735e0f1fd52f03a17d7e9b94a66f310ddfa1 [test] Add test to check mempool consistency in case of reorgs (2/2)

I didn't look closely at the test yet, since it seems pretty complex. If you could a write a one or two sentence comment before the test describing how it works, that might be helpful.

I also think it would be useful to have a comment on the `mempool::cs` variable. It's definitely being used in a way I wouldn't have expected. From first appearance, it looks like it's just being used to protect `mapTx` and allow maybe reads & writes from multiple threads:

https://github.com/bitcoin/bitcoin/blob/5d605b27457dd307e31ac320c83f3b01a41e1ae1/src/txmempool.h#L488-L489

But apparently it's supposed to not just provide thread safety, but also guarantee consistency between `mempool` and `chainActive`? Would suggest maybe:

```c++
    /**
     * Lock to guarantee consistency between mempool and external chain state.
     *
     * Lock must be held whenever mempool is read from or written to and
     * whenever a block is connected. Lock can be released when the mempool
     * contains no transactions that conflict internally or externally. It
     * should be acquired after `cs_main` if both locks need to be held at the
     * same time.
     */
    mutable CCriticalSection cs;
```

re: https://github.com/bitcoin/bitcoin/pull/14193#issuecomment-421375559

> I think it'd be good to have some documentation (whether a project document or even a gist explaining our thinking) explaining the consistency guarantees for RPC calls with respect to the mempool, validation, and the wallet

Documentation was added in #14592.

File: src/rpc/mining.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/rpc/mining.cpp]: > It seems like there is no difference in behavior here except to avoid holding g_best_block_mutex while calling mempool.GetTransactionsUpdated()?

That is indeed the motivation and required to avoid a deadlock. (You should be able to see the deadlock stack by reverting this hunk and running the functional test suite)

File: src/validation.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/validation.cpp]: It is held just as long as cs_main, because anything shorter would allow other threads (potentially rpc threads) to read from the mempool and get a result that is most likely useless in case we are in the middle of a larger reorg.

File: src/rpc/mining.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Oh, and the wallet lock was removed a long time ago, but the comment was never updated.

File: src/validation.cpp [hunk 3]
```diff
@@ -2794,6 +2795,7 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
 bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& chainparams, CBlockIndex *pindex)
 {
     AssertLockHeld(cs_main);
+    LOCK(::mempool.cs);
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/validation.cpp]: Calling it right before UpdateMempoolForReorg is too fragile and exactly the issue I want to avoid. (E.g. we `DisconnetTip` and then fail to call `UpdateMempoolForReorg`, because another thread acquired the lock)

At the very least the mempool lock should be held as long as disconnectpool sits on the stack.

**Comment by MarcoFalke** [comment]
The test is trivial, but looks scary because of the boilerplate overhead of the cpp unit tests. Pretty much the only thing it is doing is to check that during a reorg a (simulated) rpc thread (e.g. getrawmempool) would either return a consistent mempool as of before the reorg or a consistent mempool as of after the reorg and never something inconsistent in the middle of a reorg.

**Comment by MarcoFalke** [comment]
Added a commit with further comments and documentation.

File: src/rpc/mining.cpp (hunk 2 above)
Comment (jamesob) [reply src/rpc/mining.cpp]: This hunk as-is seems a little overcomplicated to me too if all we want to do is release `g_best_block_mutex` for the mempool operation. This patch (atop `master`) should do:
```patch
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index c0287ec17..b7f324a97 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -514,9 +514,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             {
                 if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
                 {
-                    // Timeout: Check transactions for update
+                    // Drop the best block mutex while accessing the mempool to avoid deadlock with
+                    // ::mempool.cs.
+                    LEAVE_CRITICAL_SECTION(lock);
+                    // Break if we've seen new transactions after the timeout has elapsed.
                     if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
                         break;
+                    ENTER_CRITICAL_SECTION(lock);
+
                     checktxtime += std::chrono::seconds(10);
                 }
             }
```

**Review by jamesob — COMMENTED** [review]
The change in general seems fine to me (mempool state should update atomically after a contiguous batch of chain maintenance and not during) and the test is great, but it might be nice to:

- explain what the motivations are. There's an argument for RPC client consistency, but IIRC this may also be a prereq for Dandelion?
- add comments for new lock acquisitions explaining why we're grabbing a seemingly unrelated lock at a coarse grain (in line with what @ryanofsky is saying).

In practice, `ActivateBestChain` calls are very short and so I don't think there are any real-world downsides to locking `::mempool.cs` for their duration - although I'm curious if this affects mempool access during IBD/reindex.

**Comment by MarcoFalke** [comment]
> add comments for new lock acquisitions explaining why we're grabbing a seemingly unrelated lock at a coarse grain (in line with what @ryanofsky is saying).

I think I already did that. Let me know if any of them are still unclear.

File: src/txmempool.h [hunk 4]
```diff
@@ -559,7 +559,7 @@ class CTxMemPool
      * Check that none of this transactions inputs are in the mempool, and thus
      * the tx is not dependent on other mempool transactions to be included in a block.
      */
-    bool HasNoInputsOf(const CTransaction& tx) const;
+    bool HasNoInputsOf(const CTransaction& tx) const EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Comment (jamesob) [inline src/txmempool.h]: Would it be clearer to just add `LOCK(cs);` to the top of [the function definition](https://github.com/jamesob/bitcoin/blob/81bd349c9c8dd247d32ece1adcc6db1254ef8971/src/txmempool.cpp#L885-L886) instead of this annotation since there's a recursive lock acquisition in `CTxMempool::exists` anyway?

File: src/validation.cpp [hunk 5]
```diff
@@ -436,7 +436,7 @@ bool CheckSequenceLocks(const CTxMemPool& pool, const CTransaction& tx, int flag
 // Returns the script flags which should be checked for a given block
 static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consensus::Params& chainparams);
 
-static void LimitMempoolSize(CTxMemPool& pool, size_t limit, unsigned long age) {
+static void LimitMempoolSize(CTxMemPool& pool, size_t limit, unsigned long age) EXCLUSIVE_LOCKS_REQUIRED(pool.cs) {
```
Comment (jamesob) [inline src/validation.cpp]: We require the `pool.cs` lock for the duration of this function (instead of just relying on the lock in `pool.Expire`) because we want to keep the mempool from changing while `pcoinsTip` modifications happen, correct?

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/14193/commits/fae71ffc99b04bbcb96f4e2b6fa25ed2fe258912

Just a few clarifying questions.

File: src/txmempool.h (hunk 4 above)
Comment (MarcoFalke) [reply src/txmempool.h]: Imo it is clearer to let the caller figure out how long the lock is required to be held. Imagine this is called in a loop, the you, most likely, wouldn't want to release the lock in between each iteration? Also, it happens that all callers already have the lock taken, since this annotation can be added without any other changes.

File: src/validation.cpp (hunk 5 above)
Comment (MarcoFalke) [reply src/validation.cpp]: Yeah, calling `LimitMempoolSize` only makes sense after you have written to the mempool, in which case you already have acquired the lock and can keep it for `LimitMempoolSize`.

File: src/txmempool.h (hunk 4 above)
Comment (jamesob) [reply src/txmempool.h]: Yep, makes sense.

File: src/rpc/mining.cpp [hunk 6]
```diff
@@ -478,9 +478,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
...
                 if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
                 {
-                    // Timeout: Check transactions for update
+                    // Drop the best block mutex while accessing the mempool to avoid deadlock with
+                    // ::mempool.cs.
+                    LEAVE_CRITICAL_SECTION(lock);
+                    // Break if we've seen new transactions after the timeout has elapsed.
                     if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
                         break;
+                    ENTER_CRITICAL_SECTION(lock);
```
Comment (practicalswift) [inline src/rpc/mining.cpp]: `ENTER_CRITICAL_SECTION(…);` isn't RAII, right?

Do we need to `LEAVE_CRITICAL_SECTION(…);` when exiting the loop?

Comment (jamesob) [reply src/rpc/mining.cpp]: Nope; the `unlock()`/debug stack pop that `LEAVE_CRITICAL_SECTION(...)` would perform is implicit to the destruction of `lock` when it falls out of scope after the `break`.

Comment (practicalswift) [reply src/rpc/mining.cpp]: @jamesob Makes perfect sense: saw the `WAIT_LOCK(g_best_block_mutex, lock);` now. Thanks.

**Review by ryanofsky — COMMENTED** [review]
I requested this in my last review (https://github.com/bitcoin/bitcoin/pull/14193#pullrequestreview-174491669) but I really think [`mempool.cs`](https://github.com/bitcoin/bitcoin/blob/5f23460c7e316fe7c944680f3feff07ebb867f70/src/txmempool.h#L488-L489) needs a comment saying how it's supposed to be used and what invariants it guarantees. Here's an attempt at a comment based on what I currently understand about what you're trying to do with this PR:

```c++
//! Recursive mutex ensuring internal consistency of transactions within the
//! mempool, and external consistency between transactions in the mempool and
//! the current chain tip. This mutex must be locked:
//!
//! - whenever reading or writing to `mempool::mapTx`
//! - whenever writing to `chainActive` or `pcoinsTip`
//!
//! and be held as long as mempool state and chain state are not consistent with
//! each other.
//!
//! Locking either the `cs_main` mutex or the `mempool::cs` mutex will lock the
//! chain tip and prevent it from being updated. But having two mutexes does
//! allow some useful concurrency, because it lets a thread that reads or writes
//! to the mempool run at the same time as a thread that reads the chain state.
mutable CCriticalSection cs;
```

**Comment by ryanofsky** [comment]
I'm probably missing some context, but what is wrong with a simpler design where you just acquire `cs_main` to get a consistent view of chain state, and `mempool::cs` to get a consistent view of the mempool, and both to get a consistent view of both together?

The new unit test and the getrawmempool example both seem very contrived. Why do these cases need to only acquire `mempool::cs` without acquiring `cs_main`? If they need to views of the mempool consistent with the current chain tip, why don't they just acquire both mutexes?

**Comment by ryanofsky** [comment]
It looks like #14963 replaces this PR. Perhaps this should be closed, or marked dependent.

**Comment by ryanofsky** [comment]
> It looks like #14963 replaces this PR. Perhaps this should be closed, or marked dependent.

Turns out #14963 doesn't really replace this PR, it just documents a bug which this PR (which is currently WIP) will fix:

https://github.com/bitcoin/bitcoin/blob/fa5c346c5a336ccdce8e50befb53746c280f053e/src/txmempool.h#L515-L522

> I'm probably missing some context, but what is wrong with a simpler design where you just acquire cs_main to get a consistent view of chain state, and mempool::cs to get a consistent view of the mempool, and both to get a consistent view of both together?

Responding to my own question, but answer from conversation seems to be that we want to allow applications that frequently query the mempool to (1) only acquire cs.mempool without cs_main for better performance and (2) not see any missing transactions if they happen to query the mempool in the middle of a reorg.

**Comment by MarcoFalke** [comment]
Addressed some of @ryanofsky's feedback

File: src/rpc/mining.cpp [hunk 7]
```diff
@@ -478,9 +478,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             {
                 if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
                 {
-                    // Timeout: Check transactions for update
+                    // Drop the best block mutex while accessing the mempool to avoid deadlock with
+                    // ::mempool.cs.
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Does you comment still apply after my latest rebase? Also, `ACQUIRED_AFTER` only works in very limited setttings (e.g. not across translation units or even function bodies). We don't use it anywhere else, so I will leave this as is for now.

File: src/test/validation_block_tests.cpp [hunk 8]
```diff
@@ -181,4 +197,116 @@ BOOST_AUTO_TEST_CASE(processnewblock_signals_ordering)
...
+                if (::mempool.mapTx.size() == 0) {
+                    // We are done with the reorg
+                    break;
+                }
+                // Internally, we might be in the middle of the reorg, but
+                // externally the reorg to the most-proof-of-work chain should
+                // be atomic. So the caller assumes that the returned mempool
+                // is consistent. That is, it has all txs that were there
+                // before the reorg.
+                assert(::mempool.mapTx.size() == txs.size());
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/validation_block_tests.cpp]: The test has a ton of comments. Let me know if I need to add anything else to parts where the code is not self-explanatory.

File: src/validation.cpp [hunk 9]
```diff
@@ -2794,6 +2795,7 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
 bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& chainparams, CBlockIndex *pindex)
 {
     AssertLockHeld(cs_main);
+    LOCK(::mempool.cs); // Lock for as long as disconnectpool is in scope to make sure UpdateMempoolForReorg is called after DisconnetTip without unlocking in between
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/validation.cpp]: Added locking annotations.

**Comment by MarcoFalke** [comment]
Anything left to do here?

File: src/txmempool.h [hunk 10]
```diff
@@ -511,15 +511,6 @@ class CTxMemPool
      * `mempool.cs` whenever adding transactions to the mempool and whenever
      * changing the chain tip. It's necessary to keep both mutexes locked until
      * the mempool is consistent with the new chain tip and fully populated.
-     *
-     * @par Consistency bug
-     *
-     * The second guarantee above is not currently enforced, but
-     * https://github.com/bitcoin/bitcoin/pull/14193 will fix it. No known code
-     * in bitcoin currently depends on second guarantee, but it is important to
-     * fix for third party code that needs be able to frequently poll the
```
Comment (ryanofsky) [inline src/txmempool.h]: In commit "validation: Add missing mempool locks" (fadeeadd93a928c8ececa8c96ad2afa45bd25a29)

Might be worth keeping text that says why guarantee 2 is useful. James also asked about it here https://github.com/bitcoin/bitcoin/pull/14193#pullrequestreview-179871886, wondering if it might be useful for dandelion.

**Review by ryanofsky — APPROVED** [review]
utACK fa33e8f41fe14cfd136ac7c96e192a9e14f1d0d3. Looks good! This makes a lot more sense to me now, and the test is actually pretty clear too.

File: src/txmempool.h (hunk 10 above)
Comment (MarcoFalke) [reply src/txmempool.h]: Not sure about that. We generally don't update (forget to update) comments when adding new features. So if a feature is added in the far future (e.g. Dandelion has no chance to hit any time soon), the phrase "No known code in bitcoin currently depends on second guarantee" might no longer reflect reality.

I believe the test that I added serves a a nice documentation that if this guarantee is violated, it will fail. The motivation to fix this bug for (enterprise) rpc users that poll the mempool should be enough motivation on its own.

File: src/validation.cpp [hunk 11]
```diff
@@ -2793,6 +2794,7 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
         LimitValidationInterfaceQueue();
 
         LOCK(cs_main);
+        LOCK(::mempool.cs); // Lock for as long as disconnectpool is in scope to make sure UpdateMempoolForReorg is called after DisconnetTip without unlocking in between
```
Comment (practicalswift) [inline src/validation.cpp]: Nit: Should be "DisconnectTip" :-)

**Review by CLEANSKY97 — COMMENTED** [review]
تکراری از #

**Review by ryanofsky — APPROVED** [review]
utACK fa81e866ebb2533489cea13618af0d58a1a355fd. Only change since last review is rebase after #15788 and fixing DisconnetTip spelling in comment

**Comment by MarcoFalke** [comment]
rewritten test now that segwit is always active in unit tests:

* Add missing call to `GenerateCoinbaseCommitment`
* Mine the reorg before adding txs to mempool, to not include them in the reorg blocks

**Review by ryanofsky — APPROVED** [review]
utACK faa0d61f8ab42cc921bdcd73d6fd4b8b680aa22e. Only changes since last review are rebase and described test changes.

> Mine the reorg before adding txs to mempool, to not include them in the reorg blocks.

I don't understand how it'd be possible for transactions in the mempool to end up in the reorg, if the reorg code is just calling `Block()` and generating empty blocks. I also don't understand what changed in the rebase if mining the reorg after adding transactions used to work, but now stopped working. In any case, it'd be useful to have a comment in the test saying that it's important to mine the reorg before adding transactions, and why, assuming it is important.

**Comment by MarcoFalke** [comment]
The mining is using CreateNewBlock, which gets txs from the mempool (if available).

Why this worked previously, IIRC, was that segwit wasn't active on regtest in unit tests, but non-standard txs would be accepted to the mempool. (See https://github.com/bitcoin/bitcoin/pull/15788/files#diff-01edb0ae76ec7f6061ec6ddbdb18375aL45 and #15891). Since the txs have witness, they wouldn't be mined previously, but now that segwit is always active on regtest, they are.

In this test we don't want to mine any txs from the mempool, so we mine the reorg when the mempool is empty and then add the txs to the mempool.

**Comment by MarcoFalke** [comment]
Also, the test is already clear on this: `// Mine a reorg (and hold it back) before adding the txs to the mempool`. If you have specific concerns, please suggest a rewording of that sentence.

**Review by CLEANSKY97 — APPROVED** [review]

File: src/rpc/mining.cpp [hunk 12]
```diff
@@ -479,9 +479,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
...
+                    LEAVE_CRITICAL_SECTION(lock);
```
Comment (sipa) [inline src/rpc/mining.cpp]: This is a pretty weird construction; generally you shouldn't be releasing the lock that protects the state you're waiting to change while inspecting it.

What about making `CTxMemPool::nTransactionsUpdated` an atomic, and making `CTxMemPool::GetTransactionsUpdated` not grap `cs`? Would it still be needed to release `lock` here then?

Would you be able to write this as `g_best_block_cv.wait_until(lock, checktxtime, [&](){ return g_best_block != hashWatchedChain || !IsRPCRunning() || mempool.GetTransactionsUpdate() != nTransactionsUpdatedLastLP; });`?

**Comment by MarcoFalke** [comment]
Rebased

File: src/rpc/mining.cpp (hunk 12 above)
Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Made it atomic, but I left the refactoring change for a future pull request.

**Review by ryanofsky — APPROVED** [review]
utACK fa2b083c3feb0522baf652045efa6b73458761a3 [EDIT: was ~e284e422e75189794e24fe482819d8b1407857c3~, from bad copy and paste]. Changes since last review: rebase after #15976, adding vTxHashes lock annotation, adding new commit dropping mempool lock for nTransactionsUpdated and making it atomic to avoid deadlock between mempool lock and g_best_block_mutex

File: src/test/validation_block_tests.cpp [hunk 13]
```diff
@@ -185,4 +204,131 @@ BOOST_AUTO_TEST_CASE(processnewblock_signals_ordering)
...
+                continue;
```
Comment (promag) [inline src/test/validation_block_tests.cpp]: fa2b083c3feb0522baf652045efa6b73458761a3

Why `continue`?

File: src/test/validation_block_tests.cpp [hunk 14]
```diff
@@ -185,4 +204,131 @@ BOOST_AUTO_TEST_CASE(processnewblock_signals_ordering)
...
+            BOOST_CHECK_EQUAL(::mempool.mapTx.size(), txs.size());
+        }
+
+        // Run a thread that simulates an RPC caller that is polling while
+        // validation is doing a reorg
+        std::thread rpc_thread{[&]() {
+            // This thread is checking that the mempool either contains all of
+            // the transactions invalidated by the reorg, or none of them, and
+            // not some intermediate amount.
+            while (true) {
```
Comment (promag) [inline src/test/validation_block_tests.cpp]: fa2b083c3feb0522baf652045efa6b73458761a3

I think this is flawless, what guarantees that there's a non-atomic change once `mempool.cs` is released?

**Review by promag — COMMENTED** [review]
Concept ACK, some comments.

File: src/test/validation_block_tests.cpp (hunk 14 above)
Comment (MarcoFalke) [reply src/test/validation_block_tests.cpp]: The code changes in this pull request should guarantee that this doesn't happen here in the test (you can check it by running the test before the code changes and after), as well as for "real" rpc polling threads. Though, I didn't write a functional test for this.

File: src/test/validation_block_tests.cpp (hunk 13 above)
Comment (MarcoFalke) [reply src/test/validation_block_tests.cpp]: Because the thread should continue and not exit (for now)

Comment (promag) [reply src/test/validation_block_tests.cpp]: Why would it exit?

Comment (MarcoFalke) [reply src/test/validation_block_tests.cpp]: It would exit when it `break`s. I think this is well documented in the test case and I am not sure what you are asking exactly.

If you are asking about cpp syntax, I find cppreference really helpful. E.g. https://en.cppreference.com/w/cpp/language/break

If you are asking about this test case logic, please write a more verbose question.

**Comment by laanwj** [comment]
code review ACK fa2b083c3feb0522baf652045efa6b73458761a3

@ryanofsky didn't you ack the wrong commit above ? e284e422e75189794e24fe482819d8b1407857c3 isn't part of this PR

File: src/test/validation_block_tests.cpp (hunk 13 above)
Comment (ryanofsky) [reply src/test/validation_block_tests.cpp]: I don't get this either. If you have a loop, and an unconditional continue is the last statement in the loop, the continue isn't doing anything and could be removed.

https://github.com/bitcoin/bitcoin/blob/fa2b083c3feb0522baf652045efa6b73458761a3/src/test/validation_block_tests.cpp#L316

**Comment by ryanofsky** [comment]
> @ryanofsky didn't you ack the wrong commit above

Sorry, bad copy and paste from the command line. Edited https://github.com/bitcoin/bitcoin/pull/14193#pullrequestreview-255403869 to fix.

**Comment by laanwj** [comment]
Thanks!

File: src/test/validation_block_tests.cpp (hunk 13 above)
Comment (MarcoFalke) [reply src/test/validation_block_tests.cpp]: Oh, I now I understand that you were asking why I put a *redundant* `continue` there.

That was intentional, because I felt it was more clear back when I wrote the test. Happy to remove, but I think we shouldn't spend too much time on style of tests.

Comment (promag) [reply src/test/validation_block_tests.cpp]: I didn't mean to discuss style! I honestly though something was wrong because of the pointless `continue`, like missing code or something else.

Comment (MarcoFalke) [reply src/test/validation_block_tests.cpp]: Ah, good point. Sorry for the confusion.
